File: shiva_erp/shiva_business_erp/page/stock_ledger_dashboard/stock_ledger_dashboard.js
// Copyright (c) 2025, Gopalakrishna Reddy Gogulamudi and contributors
// For license information, please see license.txt

frappe.pages['stock-ledger-dashboard'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Stock Ledger Dashboard',
		single_column: true
	});

	frappe.stock_ledger_dashboard = new StockLedgerDashboard(page);
}

class StockLedgerDashboard {
	constructor(page) {
		this.page = page;
		this.filters = {};
		this.setup_page();
		this.setup_filters();
		this.load_dashboard();
	}

	setup_page() {
		this.page.add_inner_button(__('Refresh'), () => {
			this.load_dashboard();
		});

		this.page.add_inner_button(__('Export Data'), () => {
			this.export_dashboard_data();
		});

		// Create main container
		this.$container = $('<div class="stock-ledger-dashboard"></div>').appendTo(this.page.main);
	}

	setup_filters() {
		// Date range filter
		this.page.add_field({
			fieldname: 'from_date',
			label: __('From Date'),
			fieldtype: 'Date',
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			change: () => this.load_dashboard()
		});

		this.page.add_field({
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			default: frappe.datetime.get_today(),
			change: () => this.load_dashboard()
		});

		// Item filter
		this.page.add_field({
			fieldname: 'item_code',
			label: __('Item'),
			fieldtype: 'Link',
			options: 'Item',
			change: () => this.load_dashboard()
		});

		// Warehouse filter
		this.page.add_field({
			fieldname: 'warehouse',
			label: __('Warehouse'),
			fieldtype: 'Link',
			options: 'Warehouse',
			change: () => this.load_dashboard()
		});
	}

	load_dashboard() {
		// Get filter values
		this.filters = {
			from_date: this.page.fields_dict.from_date.get_value(),
			to_date: this.page.fields_dict.to_date.get_value(),
			item_code: this.page.fields_dict.item_code.get_value(),
			warehouse: this.page.fields_dict.warehouse.get_value()
		};

		// Load data
		frappe.call({
			method: 'shiva_erp.shiva_business_erp.page.stock_ledger_dashboard.stock_ledger_dashboard.get_dashboard_data',
			args: { filters: this.filters },
			callback: (r) => {
				if (r.message) {
					this.render_dashboard(r.message);
				}
			}
		});
	}

	render_dashboard(data) {
		this.$container.empty();

		// Render summary cards
		this.render_summary_cards(data.summary);

		// Render charts
		this.render_charts(data.chart_data);

		// Render detailed tables
		this.render_detailed_tables(data.details);
	}

	render_summary_cards(summary) {
		let $cards_container = $(`
			<div class="dashboard-cards" style="display: grid; grid-template-columns: repeat(auto-fit, minmax(250px, 1fr)); gap: 20px; margin: 20px 0;">
			</div>
		`).appendTo(this.$container);

		const cards = [
			{
				title: 'Opening Stock',
				weight: summary.opening_weight,
				value: summary.opening_value,
				icon: 'fa-database',
				color: '#6c757d'
			},
			{
				title: 'Total IN',
				weight: summary.total_in_weight,
				value: summary.total_in_value,
				icon: 'fa-arrow-down',
				color: '#28a745'
			},
			{
				title: 'Total OUT',
				weight: summary.total_out_weight,
				value: summary.total_out_value,
				icon: 'fa-arrow-up',
				color: '#dc3545'
			},
			{
				title: 'Closing Stock',
				weight: summary.closing_weight,
				value: summary.closing_value,
				icon: 'fa-check-circle',
				color: '#007bff'
			},
			{
				title: 'Net Movement',
				weight: summary.net_weight,
				value: summary.net_value,
				icon: 'fa-exchange',
				color: summary.net_weight >= 0 ? '#28a745' : '#dc3545'
			}
		];

		cards.forEach(card => {
			let $card = $(`
				<div class="summary-card" style="background: white; border-radius: 8px; padding: 20px; box-shadow: 0 2px 8px rgba(0,0,0,0.1); border-left: 4px solid ${card.color};">
					<div style="display: flex; justify-content: space-between; align-items: center;">
						<div>
							<div style="color: #6c757d; font-size: 14px; margin-bottom: 8px;">${card.title}</div>
							<div style="font-size: 28px; font-weight: bold; color: ${card.color};">${card.weight.toFixed(2)} <span style="font-size: 16px;">Kg</span></div>
							<div style="font-size: 20px; font-weight: 600; color: ${card.color}; margin-top: 5px;">₹ ${card.value.toFixed(2)}</div>
						</div>
						<div>
							<i class="fa ${card.icon}" style="font-size: 48px; color: ${card.color}; opacity: 0.2;"></i>
						</div>
					</div>
				</div>
			`).appendTo($cards_container);
		});
	}

	render_charts(chart_data) {
		let $charts_container = $(`
			<div class="dashboard-charts" style="display: grid; grid-template-columns: repeat(auto-fit, minmax(500px, 1fr)); gap: 20px; margin: 20px 0;">
			</div>
		`).appendTo(this.$container);

		// Daily IN/OUT Weight Chart
		this.render_chart($charts_container, {
			title: 'Daily Stock Movement (Weight)',
			data: chart_data.daily_movement_weight,
			type: 'bar'
		});

		// Daily IN/OUT Value Chart
		this.render_chart($charts_container, {
			title: 'Daily Stock Movement (Value)',
			data: chart_data.daily_movement_value,
			type: 'bar'
		});

		// Balance Trend Chart (Weight and Value)
		this.render_chart($charts_container, {
			title: 'Stock Balance Trend (Weight & Value)',
			data: chart_data.balance_trend,
			type: 'line'
		});

		// Item-wise Distribution (if multiple items)
		if (chart_data.item_distribution && chart_data.item_distribution.labels.length > 1) {
			this.render_chart($charts_container, {
				title: 'Item-wise Distribution (by Weight)',
				data: chart_data.item_distribution,
				type: 'pie'
			});
		}
	}

	render_chart($container, config) {
		let $chart_card = $(`
			<div class="chart-card" style="background: white; border-radius: 8px; padding: 20px; box-shadow: 0 2px 8px rgba(0,0,0,0.1);">
				<h4 style="margin-bottom: 15px; color: #333;">${config.title}</h4>
				<div class="chart-container"></div>
			</div>
		`).appendTo($container);

		new frappe.Chart($chart_card.find('.chart-container')[0], {
			data: config.data,
			type: config.type,
			height: 250,
			colors: ['#28a745', '#dc3545', '#007bff', '#ffc107', '#6f42c1']
		});
	}

	render_detailed_tables(details) {
		let $tables_container = $(`
			<div class="dashboard-tables" style="margin: 20px 0;">
			</div>
		`).appendTo(this.$container);

		// Top items by weight movement
		if (details.top_items && details.top_items.length > 0) {
			this.render_table($tables_container, 'Top Items by Weight & Value Movement', details.top_items, [
				{ label: 'Item', field: 'item_code' },
				{ label: 'Transactions', field: 'transaction_count' },
				{ label: 'Total Weight (Kg)', field: 'total_weight' },
				{ label: 'Total Value (₹)', field: 'total_value' }
			]);
		}

		// Warehouse-wise summary
		if (details.warehouse_summary && details.warehouse_summary.length > 0) {
			this.render_table($tables_container, 'Warehouse-wise Summary', details.warehouse_summary, [
				{ label: 'Warehouse', field: 'warehouse' },
				{ label: 'IN Weight (Kg)', field: 'in_weight' },
				{ label: 'OUT Weight (Kg)', field: 'out_weight' },
				{ label: 'Balance Weight (Kg)', field: 'balance_weight' },
				{ label: 'IN Value (₹)', field: 'in_value' },
				{ label: 'OUT Value (₹)', field: 'out_value' },
				{ label: 'Balance Value (₹)', field: 'balance_value' }
			]);
		}

		// Recent transactions
		if (details.recent_transactions && details.recent_transactions.length > 0) {
			this.render_table($tables_container, 'Recent Transactions', details.recent_transactions, [
				{ label: 'Date', field: 'posting_date' },
				{ label: 'Type', field: 'transaction_type' },
				{ label: 'Voucher', field: 'voucher_no' },
				{ label: 'Item', field: 'item_code' },
				{ label: 'Weight Change (Kg)', field: 'weight_change' },
				{ label: 'Rate/Kg (₹)', field: 'rate_per_kg' },
				{ label: 'Value (₹)', field: 'value_amount' }
			]);
		}
	}

	render_table($container, title, data, columns) {
		let $table_card = $(`
			<div class="table-card" style="background: white; border-radius: 8px; padding: 20px; margin-bottom: 20px; box-shadow: 0 2px 8px rgba(0,0,0,0.1);">
				<h4 style="margin-bottom: 15px; color: #333;">${title}</h4>
				<div class="table-responsive"></div>
			</div>
		`).appendTo($container);

		let $table = $('<table class="table table-bordered table-hover"></table>').appendTo($table_card.find('.table-responsive'));

		// Header
		let $thead = $('<thead style="background: #f8f9fa;"></thead>').appendTo($table);
		let $tr = $('<tr></tr>').appendTo($thead);
		columns.forEach(col => {
			$tr.append(`<th>${col.label}</th>`);
		});

		// Body
		let $tbody = $('<tbody></tbody>').appendTo($table);
		data.forEach(row => {
			let $row = $('<tr></tr>').appendTo($tbody);
			columns.forEach(col => {
				let value = row[col.field];
				if (typeof value === 'number') {
					value = value.toFixed(2);
				}
				$row.append(`<td>${value || ''}</td>`);
			});
		});
	}

	export_dashboard_data() {
		frappe.msgprint(__('Dashboard data export feature coming soon...'));
	}
}

File: shiva_erp/shiva_business_erp/page/stock_ledger_dashboard/stock_ledger_dashboard.py
# Copyright (c) 2025, Gopalakrishna Reddy Gogulamudi and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate


@frappe.whitelist()
def get_dashboard_data(filters=None):
	"""
	Get comprehensive dashboard data for stock ledger

	Args:
		filters: dict with from_date, to_date, item_code, warehouse

	Returns:
		dict with summary, chart_data, and details
	"""
	if isinstance(filters, str):
		import json

		filters = json.loads(filters)

	if not filters:
		filters = {}

	# Get ledger data
	ledger_data = get_ledger_data(filters)

	# Calculate summary metrics
	summary = calculate_summary_metrics(ledger_data, filters)

	# Prepare chart data
	chart_data = prepare_chart_data(ledger_data)

	# Get detailed breakdowns
	details = get_detailed_breakdowns(ledger_data, filters)

	return {"summary": summary, "chart_data": chart_data, "details": details}


def get_ledger_data(filters):
	"""Get all ledger entries based on filters"""
	conditions = []
	params = {}

	if filters.get("from_date"):
		conditions.append("posting_date >= %(from_date)s")
		params["from_date"] = filters["from_date"]

	if filters.get("to_date"):
		conditions.append("posting_date <= %(to_date)s")
		params["to_date"] = filters["to_date"]

	if filters.get("item_code"):
		conditions.append("item_code = %(item_code)s")
		params["item_code"] = filters["item_code"]

	if filters.get("warehouse"):
		conditions.append("warehouse = %(warehouse)s")
		params["warehouse"] = filters["warehouse"]

	where_clause = " AND ".join(conditions) if conditions else "1=1"

	query = f"""
		SELECT
			name, posting_date, transaction_type, voucher_type, voucher_no,
			item_code, warehouse, batch_no, stock_qty, weight_kg,
			qty_change, weight_change, avg_weight_per_bird,
			rate_per_kg, value_amount
		FROM
			`tabStock Weight Ledger`
		WHERE
			{where_clause}
		ORDER BY
			posting_date ASC, creation ASC
	"""

	return frappe.db.sql(query, params, as_dict=1)


def calculate_summary_metrics(ledger_data, filters):
	"""Calculate summary metrics from ledger data - Weight and Value"""
	total_in_weight = 0
	total_out_weight = 0
	total_in_value = 0
	total_out_value = 0

	for row in ledger_data:
		if row.transaction_type == "IN":
			total_in_weight += abs(flt(row.weight_change))
			total_in_value += abs(flt(row.value_amount))
		elif row.transaction_type == "OUT":
			total_out_weight += abs(flt(row.weight_change))
			total_out_value += abs(flt(row.value_amount))
		else:
			# If transaction_type is NULL/empty, determine from weight_change sign
			if flt(row.weight_change) > 0:
				total_in_weight += abs(flt(row.weight_change))
				total_in_value += abs(flt(row.value_amount))
			else:
				total_out_weight += abs(flt(row.weight_change))
				total_out_value += abs(flt(row.value_amount))

	# Get opening balance
	opening = get_opening_balance(filters)
	opening_weight = opening.get("weight", 0)
	opening_value = opening.get("value", 0)

	# Calculate closing
	closing_weight = opening_weight + total_in_weight - total_out_weight
	closing_value = opening_value + total_in_value - total_out_value

	return {
		"opening_weight": opening_weight,
		"opening_value": opening_value,
		"total_in_weight": total_in_weight,
		"total_in_value": total_in_value,
		"total_out_weight": total_out_weight,
		"total_out_value": total_out_value,
		"closing_weight": closing_weight,
		"closing_value": closing_value,
		"net_weight": total_in_weight - total_out_weight,
		"net_value": total_in_value - total_out_value,
		"total_transactions": len(ledger_data),
	}


def get_opening_balance(filters):
	"""Get opening balance before from_date"""
	conditions = []
	params = {}

	if filters.get("from_date"):
		conditions.append("posting_date < %(from_date)s")
		params["from_date"] = filters["from_date"]
	else:
		return {"qty": 0, "weight": 0}

	if filters.get("item_code"):
		conditions.append("item_code = %(item_code)s")
		params["item_code"] = filters["item_code"]

	if filters.get("warehouse"):
		conditions.append("warehouse = %(warehouse)s")
		params["warehouse"] = filters["warehouse"]

	where_clause = " AND ".join(conditions)

	query = f"""
		SELECT
			COALESCE(SUM(weight_change), 0) as opening_weight,
			COALESCE(SUM(
				CASE 
					WHEN transaction_type = 'IN' THEN value_amount
					WHEN transaction_type = 'OUT' THEN -value_amount
					ELSE 0
				END
			), 0) as opening_value
		FROM
			`tabStock Weight Ledger`
		WHERE
			{where_clause}
	"""

	result = frappe.db.sql(query, params, as_dict=1)
	return (
		{"weight": flt(result[0].opening_weight), "value": flt(result[0].opening_value)}
		if result
		else {"weight": 0, "value": 0}
	)


def prepare_chart_data(ledger_data):
	"""Prepare data for charts"""
	# Group by date
	date_wise = {}

	for row in ledger_data:
		date = str(row.posting_date)
		if date not in date_wise:
			date_wise[date] = {
				"in_weight": 0,
				"out_weight": 0,
				"balance_weight": 0,
				"in_value": 0,
				"out_value": 0,
				"balance_value": 0,
			}

		if row.transaction_type == "IN":
			date_wise[date]["in_weight"] += abs(flt(row.weight_change))
			date_wise[date]["in_value"] += abs(flt(row.value_amount))
		else:
			date_wise[date]["out_weight"] += abs(flt(row.weight_change))
			date_wise[date]["out_value"] += abs(flt(row.value_amount))

	# Calculate running balance
	sorted_dates = sorted(date_wise.keys())
	balance_weight = 0
	balance_value = 0

	for date in sorted_dates:
		balance_weight += date_wise[date]["in_weight"] - date_wise[date]["out_weight"]
		balance_value += date_wise[date]["in_value"] - date_wise[date]["out_value"]
		date_wise[date]["balance_weight"] = balance_weight
		date_wise[date]["balance_value"] = balance_value

	# Format for charts
	labels = sorted_dates
	in_weight_data = [date_wise[d]["in_weight"] for d in labels]
	out_weight_data = [date_wise[d]["out_weight"] for d in labels]
	balance_weight_data = [date_wise[d]["balance_weight"] for d in labels]
	in_value_data = [date_wise[d]["in_value"] for d in labels]
	out_value_data = [date_wise[d]["out_value"] for d in labels]
	balance_value_data = [date_wise[d]["balance_value"] for d in labels]

	# Item-wise distribution by weight and value
	item_wise_weight = {}
	item_wise_value = {}
	for row in ledger_data:
		if row.item_code not in item_wise_weight:
			item_wise_weight[row.item_code] = 0
			item_wise_value[row.item_code] = 0
		item_wise_weight[row.item_code] += abs(flt(row.weight_change))
		item_wise_value[row.item_code] += abs(flt(row.value_amount))

	return {
		"daily_movement_weight": {
			"labels": labels,
			"datasets": [
				{"name": "IN (Kg)", "values": in_weight_data},
				{"name": "OUT (Kg)", "values": out_weight_data},
			],
		},
		"daily_movement_value": {
			"labels": labels,
			"datasets": [
				{"name": "IN (₹)", "values": in_value_data},
				{"name": "OUT (₹)", "values": out_value_data},
			],
		},
		"balance_trend": {
			"labels": labels,
			"datasets": [
				{"name": "Weight Balance (Kg)", "values": balance_weight_data},
				{"name": "Value Balance (₹)", "values": balance_value_data},
			],
		},
		"item_distribution": {
			"labels": list(item_wise_weight.keys()),
			"datasets": [
				{"name": "Weight (Kg)", "values": list(item_wise_weight.values())},
				{"name": "Value (₹)", "values": list(item_wise_value.values())},
			],
		},
	}


def get_detailed_breakdowns(ledger_data, filters):
	"""Get detailed breakdowns for tables"""
	# Top items by weight and value movement
	item_stats = {}
	for row in ledger_data:
		if row.item_code not in item_stats:
			item_stats[row.item_code] = {
				"item_code": row.item_code,
				"transaction_count": 0,
				"total_weight": 0,
				"total_value": 0,
			}

		item_stats[row.item_code]["transaction_count"] += 1
		item_stats[row.item_code]["total_weight"] += abs(flt(row.weight_change))
		item_stats[row.item_code]["total_value"] += abs(flt(row.value_amount))

	top_items = sorted(item_stats.values(), key=lambda x: x["total_weight"], reverse=True)[:10]

	# Warehouse-wise summary
	warehouse_stats = {}
	for row in ledger_data:
		if row.warehouse not in warehouse_stats:
			warehouse_stats[row.warehouse] = {
				"warehouse": row.warehouse,
				"in_weight": 0,
				"out_weight": 0,
				"balance_weight": 0,
				"in_value": 0,
				"out_value": 0,
				"balance_value": 0,
			}

		if row.transaction_type == "IN":
			warehouse_stats[row.warehouse]["in_weight"] += abs(flt(row.weight_change))
			warehouse_stats[row.warehouse]["in_value"] += abs(flt(row.value_amount))
		else:
			warehouse_stats[row.warehouse]["out_weight"] += abs(flt(row.weight_change))
			warehouse_stats[row.warehouse]["out_value"] += abs(flt(row.value_amount))

	# Get current balance for each warehouse
	for wh in warehouse_stats:
		balance = get_current_balance(filters.get("item_code"), wh, filters.get("to_date"))
		warehouse_stats[wh]["balance_weight"] = balance.get("weight", 0)
		warehouse_stats[wh]["balance_value"] = balance.get("value", 0)

	# Recent transactions
	recent_transactions = ledger_data[-20:] if len(ledger_data) > 20 else ledger_data
	recent_transactions.reverse()

	return {
		"top_items": top_items,
		"warehouse_summary": list(warehouse_stats.values()),
		"recent_transactions": recent_transactions,
	}


def get_current_balance(item_code, warehouse, to_date=None):
	"""Get current balance for item-warehouse"""
	conditions = []
	params = {}

	if item_code:
		conditions.append("item_code = %(item_code)s")
		params["item_code"] = item_code

	if warehouse:
		conditions.append("warehouse = %(warehouse)s")
		params["warehouse"] = warehouse

	if to_date:
		conditions.append("posting_date <= %(to_date)s")
		params["to_date"] = to_date

	where_clause = " AND ".join(conditions) if conditions else "1=1"

	query = f"""
		SELECT
			COALESCE(SUM(weight_change), 0) as balance_weight,
			COALESCE(SUM(
				CASE 
					WHEN transaction_type = 'IN' THEN value_amount
					WHEN transaction_type = 'OUT' THEN -value_amount
					ELSE 0
				END
			), 0) as balance_value
		FROM
			`tabStock Weight Ledger`
		WHERE
			{where_clause}
	"""

	result = frappe.db.sql(query, params, as_dict=1)
	return (
		{"weight": flt(result[0].balance_weight), "value": flt(result[0].balance_value)}
		if result
		else {"weight": 0, "value": 0}
	)
